diff --git a/dex2oat/dex2oat.cc b/dex2oat/dex2oat.cc
index 729d712a54..5529f0a2d2 100644
--- a/dex2oat/dex2oat.cc
+++ b/dex2oat/dex2oat.cc
@@ -1035,6 +1035,24 @@ class Dex2Oat FINAL {
                                                         : OatHeader::kFalseValue);
   }
 
+  //patch by Youlor
+  //++++++++++++++++++++++++++++
+  const char* UNPACK_CONFIG = "/data/local/tmp/unpacker.config";
+  bool ShouldUnpack() {
+    std::ifstream config(UNPACK_CONFIG);
+    std::string line;
+    if(config) {
+        while (std::getline(config, line)) { 
+          std::string package_name = line.substr(0, line.find(':'));
+          if (oat_location_.find(package_name) != std::string::npos) {
+            return true;
+          }
+        }
+    }
+    return false;
+  }
+  //++++++++++++++++++++++++++++
+
   // Parse the arguments from the command line. In case of an unrecognized option or impossible
   // values/combinations, a usage error will be displayed and exit() is called. Thus, if the method
   // returns, arguments have been successfully parsed.
@@ -1179,6 +1197,13 @@ class Dex2Oat FINAL {
 
     // Insert some compiler things.
     InsertCompileOptions(argc, argv);
+
+    //patch by Youlor
+    //++++++++++++++++++++++++++++
+    if (ShouldUnpack()) {
+      compiler_options_->SetCompilerFilter(CompilerFilter::kVerifyAtRuntime);
+    }
+    //++++++++++++++++++++++++++++
   }
 
   // Check whether the oat output files are writable, and open them for later. Also open a swap
diff --git a/runtime/Android.mk b/runtime/Android.mk
index aa12c83ceb..fbf97f785a 100644
--- a/runtime/Android.mk
+++ b/runtime/Android.mk
@@ -193,7 +193,15 @@ LIBART_COMMON_SRC_FILES := \
   verifier/reg_type_cache.cc \
   verifier/register_line.cc \
   well_known_classes.cc \
-  zip_archive.cc
+  zip_archive.cc \
+  unpacker/unpacker.cc \
+  unpacker/cJSON.cc
+#patch by Youlor
+#++++++++++++++++++++++++++++
+  # zip_archive.cc \
+  # unpacker/unpacker.cc \
+  # unpacker/cJSON.cc
+#++++++++++++++++++++++++++++
 
 LIBART_COMMON_SRC_FILES += \
   arch/context.cc \
diff --git a/runtime/art_method.cc b/runtime/art_method.cc
index f86cb13512..7a60a9ac67 100644
--- a/runtime/art_method.cc
+++ b/runtime/art_method.cc
@@ -40,6 +40,11 @@
 #include "scoped_thread_state_change.h"
 #include "well_known_classes.h"
 
+//patch by Youlor
+//++++++++++++++++++++++++++++
+#include "unpacker/unpacker.h"
+//++++++++++++++++++++++++++++
+
 namespace art {
 
 extern "C" void art_quick_invoke_stub(ArtMethod*, uint32_t*, uint32_t, Thread*, JValue*,
@@ -255,7 +260,12 @@ void ArtMethod::Invoke(Thread* self, uint32_t* args, uint32_t args_size, JValue*
   // If the runtime is not yet started or it is required by the debugger, then perform the
   // Invocation by the interpreter, explicitly forcing interpretation over JIT to prevent
   // cycling around the various JIT/Interpreter methods that handle method invocation.
-  if (UNLIKELY(!runtime->IsStarted() || Dbg::IsForcedInterpreterNeededForCalling(self, this))) {
+  //patch by Youlor
+  //++++++++++++++++++++++++++++
+  //如果是主动调用fake invoke并且不是native方法则强制走解释器
+  if (UNLIKELY(!runtime->IsStarted() || Dbg::IsForcedInterpreterNeededForCalling(self, this) 
+      || (Unpacker::isFakeInvoke(self, this) && !this->IsNative()))) {
+  //++++++++++++++++++++++++++++
     if (IsStatic()) {
       art::interpreter::EnterInterpreterFromInvoke(
           self, this, nullptr, args, result, /*stay_in_interpreter*/ true);
@@ -266,6 +276,15 @@ void ArtMethod::Invoke(Thread* self, uint32_t* args, uint32_t args_size, JValue*
           self, this, receiver, args + 1, result, /*stay_in_interpreter*/ true);
     }
   } else {
+    //patch by Youlor
+    //++++++++++++++++++++++++++++
+    //如果是主动调用fake invoke并且是native方法则不执行
+    if (Unpacker::isFakeInvoke(self, this) && this->IsNative()) {
+      // Pop transition.
+      self->PopManagedStackFragment(fragment);
+      return;
+    }
+    //++++++++++++++++++++++++++++
     DCHECK_EQ(runtime->GetClassLinker()->GetImagePointerSize(), sizeof(void*));
 
     constexpr bool kLogInvocationStartAndReturn = false;
diff --git a/runtime/class_linker.h b/runtime/class_linker.h
index 8aceffbdcc..5c1bffa503 100644
--- a/runtime/class_linker.h
+++ b/runtime/class_linker.h
@@ -1176,6 +1176,10 @@ class ClassLinker {
   friend class ImageWriter;  // for GetClassRoots
   friend class JniCompilerTest;  // for GetRuntimeQuickGenericJniStub
   friend class JniInternalTest;  // for GetRuntimeQuickGenericJniStub
+  //patch by Youlor
+  //++++++++++++++++++++++++++++
+  friend class Unpacker;
+  //++++++++++++++++++++++++++++
   ART_FRIEND_TEST(ClassLinkerTest, RegisterDexFileName);  // for DexLock, and RegisterDexFileLocked
   ART_FRIEND_TEST(mirror::DexCacheTest, Open);  // for AllocDexCache
   DISALLOW_COPY_AND_ASSIGN(ClassLinker);
diff --git a/runtime/interpreter/interpreter.cc b/runtime/interpreter/interpreter.cc
index 8c42b3abce..690722e99a 100644
--- a/runtime/interpreter/interpreter.cc
+++ b/runtime/interpreter/interpreter.cc
@@ -240,7 +240,11 @@ static std::ostream& operator<<(std::ostream& os, const InterpreterImplKind& rhs
   return os;
 }
 
-static constexpr InterpreterImplKind kInterpreterImplKind = kMterpImplKind;
+//patch by Youlor
+//++++++++++++++++++++++++++++
+//static constexpr InterpreterImplKind kInterpreterImplKind = kMterpImplKind;
+static constexpr InterpreterImplKind kInterpreterImplKind = kSwitchImplKind;
+//++++++++++++++++++++++++++++
 
 #if defined(__clang__)
 // Clang 3.4 fails to build the goto interpreter implementation.
diff --git a/runtime/interpreter/interpreter_switch_impl.cc b/runtime/interpreter/interpreter_switch_impl.cc
index 18330babe0..b642ca9416 100644
--- a/runtime/interpreter/interpreter_switch_impl.cc
+++ b/runtime/interpreter/interpreter_switch_impl.cc
@@ -19,7 +19,10 @@
 #include "interpreter_common.h"
 #include "jit/jit.h"
 #include "safe_math.h"
-
+//patch by Youlor
+//++++++++++++++++++++++++++++
+#include "unpacker/unpacker.h"
+//++++++++++++++++++++++++++++
 #include <memory>  // std::unique_ptr
 
 namespace art {
@@ -61,13 +64,22 @@ namespace interpreter {
   }
 
 // Code to run before each dex instruction.
+//patch by Youlor
+//++++++++++++++++++++++++++++
 #define PREAMBLE()                                                                              \
   do {                                                                                          \
+    inst_count++;                                                                               \
+    bool dumped = Unpacker::beforeInstructionExecute(self, shadow_frame.GetMethod(),            \
+                                                     dex_pc, inst_count);                       \
+    if (dumped) {                                                                               \
+      return JValue();                                                                          \
+    }                                                                                           \
     if (UNLIKELY(instrumentation->HasDexPcListeners())) {                                       \
       instrumentation->DexPcMovedEvent(self, shadow_frame.GetThisObject(code_item->ins_size_),  \
                                        shadow_frame.GetMethod(), dex_pc);                       \
     }                                                                                           \
   } while (false)
+//++++++++++++++++++++++++++++
 
 #define BRANCH_INSTRUMENTATION(offset)                                                         \
   do {                                                                                         \
@@ -119,6 +131,10 @@ JValue ExecuteSwitchImpl(Thread* self, const DexFile::CodeItem* code_item,
   // to keep this live for the scope of the entire function call.
   std::unique_ptr<lambda::ClosureBuilder> lambda_closure_builder;
   size_t lambda_captured_variable_index = 0;
+  //patch by Youlor
+  //++++++++++++++++++++++++++++
+  int inst_count = -1;
+  //++++++++++++++++++++++++++++
   do {
     dex_pc = inst->GetDexPc(insns);
     shadow_frame.SetDexPC(dex_pc);
@@ -2433,6 +2449,13 @@ JValue ExecuteSwitchImpl(Thread* self, const DexFile::CodeItem* code_item,
       case Instruction::UNUSED_7A:
         UnexpectedOpcode(inst, shadow_frame);
     }
+    //patch by Youlor
+    //++++++++++++++++++++++++++++
+    bool dumped = Unpacker::afterInstructionExecute(self, shadow_frame.GetMethod(), dex_pc, inst_count);
+    if (dumped) {
+      return JValue();
+    }
+    //++++++++++++++++++++++++++++
   } while (!interpret_one_instruction);
   // Record where we stopped.
   shadow_frame.SetDexPC(inst->GetDexPc(insns));
diff --git a/runtime/runtime.cc b/runtime/runtime.cc
index 6d0d6ed7fc..178eaef543 100644
--- a/runtime/runtime.cc
+++ b/runtime/runtime.cc
@@ -140,6 +140,11 @@
 #include "verifier/method_verifier.h"
 #include "well_known_classes.h"
 
+//patch by Youlor
+//++++++++++++++++++++++++++++
+#include "unpacker/unpacker.h"
+//++++++++++++++++++++++++++++
+
 namespace art {
 
 // If a signal isn't handled properly, enable a handler that attempts to dump the Java stack.
@@ -1421,6 +1426,10 @@ void Runtime::RegisterRuntimeNativeMethods(JNIEnv* env) {
   register_org_apache_harmony_dalvik_ddmc_DdmServer(env);
   register_org_apache_harmony_dalvik_ddmc_DdmVmInternal(env);
   register_sun_misc_Unsafe(env);
+  //patch by Youlor
+  //++++++++++++++++++++++++++++
+  Unpacker::register_cn_youlor_Unpacker(env);
+  //++++++++++++++++++++++++++++
 }
 
 void Runtime::DumpForSigQuit(std::ostream& os) {
